Document how content generation works (#2170)
sourcegraph/handbook: content/handbook/editing/content-generation.md

This page describes how this generation functionality is implemented in case you need to modify it.

## Data files

The data that drives the various pages is located in [/data](https://github.com/sourcegraph/handbook/tree/main/data) and is stored in the YAML format. Each contains a list of entries, with those entries containing various fields, and those fields containing strings or pointers to other elements (for example, the name of a person as a string, or a pointer for a team to a member of that team via the id.) For example, here is a sample org where the title and strategy links are strings, and the pm and em are pointers to people:

```yaml
code_graph:
  title: Code Graph
  strategy_link: /company/strategy/code-graph/index.md
  pm: jason_yavorska
  em: yink_teo
```

There are schema files for each of the data files at [/schema](https://github.com/sourcegraph/handbook/tree/main/schema) that set formats and whether different fields are required or not.

When implementing new fields/structures, avoid creating new string values for things that you can look up via id or else we will have multiple sources of truth for the same values.

## Code

> The flow for content generation starts in [/src/lib/markdownToHtml.ts](https://github.com/sourcegraph/handbook/blob/main/src/lib/markdownToHtml.ts) in the function `insertGeneratedMarkdown`, which handles detecting a `{{generator:*}}` directive and passing it on to the appropriate markdown generation function.

Author: ```suggestion
The flow for content generation starts in [/src/lib/markdownToHtml.ts](https://github.com/sourcegraph/handbook/blob/main/src/lib/markdownToHtml.ts) in the function `insertGeneratedMarkdown`, which handles detecting a `{{generator:*}}` directive and passing it on to the appropriate markdown generation function. If you add a new kind of generator you need to add the mapping here.
```

The markdown generation functions are in [/src/lib/generatedMarkdown.js](https://github.com/sourcegraph/handbook/blob/main/src/lib/generatedMarkdown.js). These functions are responsible for processing the data files and returning the markdown that will be inserted on the page. Typically these functions are iterating over objects, looking up values for other objects, in order to generate some sort of list that can be returned for embedding on the page.

> Lastly, there is [/src/scripts/validateData.mjs](https://github.com/sourcegraph/handbook/blob/main/src/scripts/validateData.mjs) which runs at build time and validates that any data that points to other data (for example, a pointer to a specific team member from a team) actually exists, and will report an error if it does not.

Author: ```suggestion
Lastly, there is [/src/scripts/validateData.mjs](https://github.com/sourcegraph/handbook/blob/main/src/scripts/validateData.mjs) which runs at build time and validates that any data that points to other data (for example, a pointer to a specific team member from a team) actually exists, and will report an error if it does not. If you add new data relationships, you should also add a check here.
```

> Lastly, there is [/src/scripts/validateData.mjs](https://github.com/sourcegraph/handbook/blob/main/src/scripts/validateData.mjs) which runs at build time and validates that any data that points to other data (for example, a pointer to a specific team member from a team) actually exists, and will report an error if it does not. If you add new data relationships, you should also add a check here.

Reviewer: What is the philosophy if validations fail? Do we skip broken relationships? Do we throw an error on the entire generated block?

Author: It returns the invalid object as JSON and fails the validation step in the build.

## Edit data shortcut for pages driven by data

You can configure a page to show a shortcut to editing it's header as follows:

```
---
data_source: [/data/code_hosts.yml, code hosts]
data_source_2: [/data/product_teams.yml, product teams]
---
```

The page supports up to two data sources (the second is optional).

> This will show, underneath "Edit this page" on the right hand side, "Edit team data". This is implemented in [/src/components/EditSection.tsx](https://github.com/sourcegraph/handbook/blob/main/src/components/EditSection.tsx).

Author: ```suggestion
This particular example will show, underneath "Edit this page" on the right hand side, "Edit code hosts data" and "Edit product teams data" (the first entry is the path, the second is what to display visually). This is implemented in [/src/components/EditSection.tsx](https://github.com/sourcegraph/handbook/blob/main/src/components/EditSection.tsx).
```
